### PR TITLE
RBUS - Support for interval-based subscriptions

### PR DESCRIPTION
**Pull request by NetajiPanigrahi**
[RDKB-46149] RBUS - Support for interval-based subscriptions
Reason for change: Added interval based subscription, 
Test Procedure: Test and verified, Run rbusIntervalProvider, rbusIntervalConsumer
                          Used below rbuscli commands for verfication
                          1. subint [Event name] [interval] - Command to subscribe with interval
                          2. unsubint [Event name] [interval] - Command to unsubscribe the with interval that was used when 
                              subscribing to this event.
Risks: Low
Priority: P1
Signed-off-by: Netaji Panigrahi <Netaji_Panigrahi@comcast.com>

### CODE REVIEW

File: src/rbus/rbus_interval_sub.c [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+        return;
+    }
+
+    sRecord* rec;
+    rec = sub_find(sub);
+    ERROR_CHECK(pthread_mutex_unlock(&gMutex));
+
+    if(rec)
+    {
+        if(rec->running)
```
Comment (karuna2git) [inline src/rbus/rbus_interval_sub.c:268]: Use `rec->mutex` before you set `rec->running` to false;

File: src/rbus/rbus_interval_sub.c [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    ERROR_CHECK(pthread_cond_init(&sub_rec->cond, &cattrib));
+    ERROR_CHECK(pthread_condattr_destroy(&cattrib));
+}
+
+static void sub_Free(void* p)
+{
+    sRecord* rec = (sRecord*)p;
+    if(rec)
+    {
+        rbusProperty_Release(rec->property);
```
Comment (karuna2git) [inline src/rbus/rbus_interval_sub.c:84]: can  u destroy rec->mutex and rec->cond also?

File: src/rbus/rbus_interval_sub.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#ifndef RBUS_INTERVAL_H
+#define RBUS_INTERVAL_H
+
+#include "rbus_subscriptions.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void AddSubscriptionRecord(rbusHandle_t handle, elementNode* propNode, rbusSubscription_t* sub);
```
Comment (karuna2git) [inline src/rbus/rbus_interval_sub.h:29]: rename the functions as,
rbusInterval_AddSubscriptionRecord()
rbusInterval_RemoveSubscriptionRecord()

File: src/rbus/rbus_interval_sub.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (karuna2git) [inline src/rbus/rbus_interval_sub.h:1]: rename the file as,
rbus_intervalsubscription.c
rbus_intervalsubscription.h

File: src/rbus/rbus.c [hunk 5]
```diff
@@ -937,13 +963,27 @@ int subscribeHandlerImpl(
...
+            if (subscription->interval && elementHasIntervalSubscriptions(node, subscription))
             {
-                RBUSLOG_INFO("%s: ValueChange %s event=%s prop=%s", __FUNCTION__, 
-                    added ? "Add" : "Remove", subscription->eventName, node->fullName);
-
+                RBUSLOG_INFO("%s: subscription with interval  %s event=%s prop=%s", __FUNCTION__,
+                        added ? "Add" : "Remove", subscription->eventName, node->fullName);
+                if(added)
+                {
+                    AddSubscriptionRecord(handle, node, subscription);
```
Comment (karuna2git) [inline src/rbus/rbus.c:972]: rename as we discussed below

File: src/rbus/rbus.c [hunk 6]
```diff
@@ -937,13 +963,27 @@ int subscribeHandlerImpl(
...
+                    break;
+                }
+                else
+                {
+                    RemoveSubscriptionRecord(handle, node, subscription);
```
Comment (karuna2git) [inline src/rbus/rbus.c:977]: rename the function

File: src/rbus/CMakeLists.txt [hunk 7]
```diff
@@ -40,6 +40,7 @@ add_library(
     rbus_subscriptions.c
     rbus_tokenchain.c
     rbus_asyncsubscribe.c
+    rbus_interval_sub.c
```
Comment (karuna2git) [inline src/rbus/CMakeLists.txt:43]: change the file name

File: src/rbus/rbus.c [hunk 8]
```diff
@@ -33,6 +33,7 @@
 #include "rbus_valuechange.h"
 #include "rbus_subscriptions.h"
 #include "rbus_asyncsubscribe.h"
+#include "rbus_interval_sub.h"
```
Comment (karuna2git) [inline src/rbus/rbus.c:36]: change the header file name

File: src/rbus/rbus.c [hunk 9]
```diff
@@ -221,19 +222,36 @@ void rbusEventSubscription_free(void* p)
     free(sub);
 }
 
-static rbusEventSubscription_t* rbusEventSubscription_find(rtVector eventSubs, char const* eventName, rbusFilter_t filter)
+static rbusEventSubscription_t* rbusEventSubscription_find(rtVector eventSubs, char const* eventName,
+        rbusFilter_t filter, int32_t* interval, uint32_t* duration)
```
Comment (karuna2git) [inline src/rbus/rbus.c:226]: no need pass pointer to interval, duration for finding it

File: src/rbus/rbus.c [hunk 10]
```diff
@@ -221,19 +222,36 @@ void rbusEventSubscription_free(void* p)
...
+        if(sub && !strcmp(sub->eventName, eventName) && !rbusFilter_Compare(sub->filter, filter)) {
+            found = true;
+            /*Compare interval */
+            if ((interval != NULL) && (sub->interval != *interval) )
+            {
+                found = false;
+
+            }
+            /*Compare duration*/
+            if ((duration != NULL) && (sub->duration != *duration))
```
Comment (karuna2git) [inline src/rbus/rbus.c:244]: no need of pointer

File: src/rbus/rbus.c [hunk 11]
```diff
@@ -4257,7 +4301,7 @@ rbusError_t rbusEvent_Unsubscribe(
 
     /*the use of rtVector is inefficient here.  I have to loop through the vector to find the sub by name, 
         then call RemoveItem, which loops through again to find the item by address to destroy */
-    sub = rbusEventSubscription_find(handleInfo->eventSubs, eventName, NULL);
+    sub = rbusEventSubscription_find(handleInfo->eventSubs, eventName, NULL, NULL, NULL);
```
Comment (karuna2git) [inline src/rbus/rbus.c:4304]: Pass, filter as NULL(already done), interval as 0 and duration as 0.

File: src/rbus/rbus.c [hunk 12]
```diff
@@ -1199,7 +1243,7 @@ static int _master_event_callback_handler(char const* sender, char const* eventN
 
     RBUSLOG_DEBUG("Received master event callback: sender=%s eventName=%s componentId=%d", sender, eventName, componentId);
 
-    subscription = rbusEventSubscription_find(handleInfo->eventSubs, eventName, filter);
+    subscription = rbusEventSubscription_find(handleInfo->eventSubs, eventName, filter, &interval, &duration);
```
Comment (karuna2git) [inline src/rbus/rbus.c:1246]: no need to pass a pointer

File: src/rbus/rbus_interval_sub.c [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+        rbusProperty_Release(property);
+        count++;
+    }
+    ERROR_CHECK(pthread_mutex_unlock(&sub_rec->mutex));
+    RBUSLOG_DEBUG("%s: stop\n", __FUNCTION__);
+    return NULL;
+}
+
+/* Add Subscription Record to global record and start subscription thread */
+void AddSubscriptionRecord(
```
[the remark this answers is not in the document]
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: updated

File: src/rbus/rbus_interval_sub.c [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+    if(!propNode)
+    {
+        RBUSLOG_ERROR("%s: propNode NULL error", __FUNCTION__);
+        return;
+    }
+
+    if(propNode->type != RBUS_ELEMENT_TYPE_PROPERTY)
+    {
+        RBUSLOG_ERROR("%s: propNode type %d error", __FUNCTION__, propNode->type);
+        return;
```
[the remark this answers is not in the document]
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: Addressed in new patch set

File: src/rbus/rbus_interval_sub.c [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+    }
+    assert(propNode->cbTable.getHandler);
```
[the remark this answers is not in the document]
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: Removed

File: src/rbus/rbus_interval_sub.c [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    if(!propNode->cbTable.getHandler)
+    {
+        RBUSLOG_ERROR("%s: propNode getHandler NULL error", __FUNCTION__);
+        return;
```
[the remark this answers is not in the document]
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: Done

File: src/rbus/rbus_interval_sub.c [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+    if(!sub_rec)
+    {
+        sub_rec = (sRecord*)rt_malloc(sizeof(sRecord));
+        sub_rec->handle = handle;
+        sub_rec->node = propNode;
+        sub_rec->sub = sub;
+        rbusProperty_Init(&sub_rec->property, propNode->fullName, NULL);
+        init_thread(sub_rec);
+
+        rtVector_PushBack(gRecord, sub_rec);
```
[the remark this answers is not in the document]
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: Done

File: src/rbus/rbus_interval_sub.c (hunk 1 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: Working fine without adding 'rec->mutex' lock

File: src/rbus/rbus_interval_sub.c (hunk 2 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.c]: Changes are done

File: src/rbus/rbus_interval_sub.h (hunk 3 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.h]: Done

File: src/rbus/rbus_interval_sub.h (hunk 4 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus_interval_sub.h]: file names are changed as per suggestion

File: src/rbus/rbus.c (hunk 5 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus.c]: done

File: src/rbus/rbus.c (hunk 6 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus.c]: done

File: src/rbus/CMakeLists.txt (hunk 7 above)
Comment (NetajiPanigrahi) [reply src/rbus/CMakeLists.txt]: done

File: src/rbus/rbus.c (hunk 9 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus.c]: code changes are done.

File: src/rbus/rbus.c (hunk 11 above)
Comment (NetajiPanigrahi) [reply src/rbus/rbus.c]: done

**Review by karuna2git — APPROVED** [review]

**Pull request merged**